tls/s2n_key_log.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <s2n.h>
+#include "tls/s2n_config.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_crypto_constants.h"
+#include "tls/s2n_quic_support.h" /* this currently holds the s2n_secret_type_t enum */
+#include "utils/s2n_blob.h"
+#include "utils/s2n_safety.h"
+
+/* hex requires 2 chars per byte */
+#define HEX_ENCODING_SIZE 2
+
+/* https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS/Key_Log_Format */
+
+S2N_RESULT s2n_key_log_hex_encode(struct s2n_stuffer *output, uint8_t *bytes, size_t len)
+{
```

Comment:
Not a blocker, but there are existing stuffer hex functions: https://github.com/awslabs/s2n/blob/main/tests/testlib/s2n_testlib.h#L30-L35 They're just only used in the tests rn. You could move them out of the test library and reuse them.

Reply:
I can open an issue for that but I don't think it should be done here #2585

Reply:
Then we should probably test this method to make sure it's doing the hex conversions right. Your current unit tests check that you're probably writing the right string, but only your manual tests verified that the secrets are actually being encoded right.

Reply:
https://github.com/awslabs/s2n/blob/f695aa1206e33d7b83e15f6dbaa46bfb2af07241/tests/unit/s2n_self_talk_key_log_test.c#L169-L188